File: Paradigms/__2000__/2169/2169.py
from typing import List

SUCESS = "Missao completada com sucesso"
FAIL = "You Are Dead"

# armas e seu dano por munição 
WEAPONS = {
    "HANDGUN":      2,
    "RED9":         3.5,
    "BLACKTAIL":    3.5,
    "MATILDA":      2,
    "HANDCANNON":   60,
    "STRIKER":      12,
    "TMP":          1.2,
    "RIFLE":        12
}

# monstros e quantidade de vida HP
MONSTERS = {
    "GANADOS":          50,
    "COBRAS":           40,
    "ZEALOT":           75,
    "COLMILLOS":        60,
    "GARRADOR":         125,
    "LASPLAGAS":        100,
    "GATLINGMAN":       150,
    "REGENERATOR":      250,
    "ELGIGANTE":        500,
    "DR.SALVADOR":      350
}


def knapsack_problem(itens: List[List[int]], capacity: int) -> int:
    
    dp = [[0] * (capacity + 1) for _ in range(len(itens) + 1)]
    
    for i in range(1, len(itens) + 1):
        curr_item_value, curr_item_weight = itens[i - 1][0], itens[i - 1][1]
        
        for j in range(capacity + 1):
            if j < curr_item_weight: dp[i][j] = dp[i - 1][j]
            else:
                remaining_capacity = j - curr_item_weight
                added_value = dp[i - 1][remaining_capacity] + curr_item_value
                dp[i][j] = max(dp[i - 1][j], added_value)
    return dp
    


while True:
    try:
    
        n_weapons = int(input())
        weapons = []
        for _ in range(n_weapons):
            weapon_name, amount_ammunation = input().split()
            damage = WEAPONS[weapon_name]
            qnt_ammunation = int(amount_ammunation)
            weapons.append([damage * qnt_ammunation, qnt_ammunation])
          
        n_monsters = int(input())  
        total_HP_monsters = 0
        for _ in range(n_monsters):
            monster, qnt = input().split()
            total_HP_monsters += MONSTERS[monster] * int(qnt)
            
        capacity_ammunation = int(input())
        
        
        dp = knapsack_problem(weapons, capacity_ammunation)
        max_damage = dp[n_weapons][capacity_ammunation]
        
        monster_HP_final = total_HP_monsters - max_damage
        
        
        print(SUCESS if monster_HP_final <= 0 else FAIL)
        print()

    except EOFError: break 

File: Paradigms/__2000__/2169/2169.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>
#include <tuple>
#include <iomanip>
#include <algorithm>

#define SUCESS "Missao completada com sucesso"
#define FAIL "You Are Dead"

typedef std::vector<std::tuple<double, int>> ITENS;

// tabela de armas e dano por municao
const std::map<std::string, double> WEAPONS = {
    {"HANDGUN",     2.0},    {"RED9",       3.5},
    {"BLACKTAIL",   3.5},    {"MATILDA",    2.0},
    {"HANDCANNON",  60.0},   {"STRIKER",    12.0},
    {"TMP",         1.2},    {"RIFLE",      12.0}
};

// tabela de inimigos e a quantidade de vida de cada
const std::map<std::string, double> ENEMIES = {
    {"GANADOS",     50},    {"COBRAS",          40},
    {"ZEALOT",      75},    {"COLMILLOS",       60},
    {"GARRADOR",    125},   {"LASPLAGAS",       100},
    {"GATLINGMAN",  150},   {"REGENERATOR",     250},
    {"ELGIGANTE",   500},   {"DR.SALVADOR",     350}
};

std::vector<std::vector<int>> KnapsackProblem(ITENS& itens, int capacity) {
    
    int rows = itens.size() + 1;
    int cols = capacity + 1;
    std::vector<std::vector<int>> dp(rows, std::vector<int>(cols, 0));

    for (int i = 1; i < rows; i++) {
        
        double curr_damage = std::get<0>(itens[i - 1]);
        int curr_ammo = std::get<1>(itens[i - 1]);

        for (int j = 0; j < cols; j++) {
            if (j < curr_ammo) dp[i][j] = dp[i - 1][j];
            else {
                int remaining = j - curr_ammo;
                int added_value = dp[i - 1][remaining] + (int)curr_damage;
                dp[i][j] = std::max(dp[i - 1][j], added_value);
            }
        }
    }
    return dp;
}

int main() {
    using namespace std;
    int n_weapons, n_enemies, amount_ammunation, amount_enemies, capacity;
    string weapon_name, enemy_name;

    while (cin >> n_weapons) {
        ITENS weapons;
        for (int i = 0; i < n_weapons; i++) {
            cin >> weapon_name >> amount_ammunation;
            double damage = WEAPONS.at(weapon_name) * (double) amount_ammunation;
            weapons.push_back({damage, amount_ammunation});
        }

        int total_enemies_HP = 0;
        cin >> n_enemies;
        for (int i = 0; i < n_enemies; i++) {
            cin >> enemy_name >> amount_enemies;
            total_enemies_HP += ENEMIES.at(enemy_name) * amount_enemies;
        }

        cin >> capacity;
        auto dp = KnapsackProblem(weapons, capacity);
        double max_damage = dp[n_weapons][capacity];

        cout << (total_enemies_HP - max_damage <= 0 ? SUCESS : FAIL) << endl << endl;
    }

    return 0;
}